libvast/src/system/index.cpp

```diff
@@ -1203,30 +1218,38 @@ index(index_actor::stateful_pointer<index_state> self,
       self->send(lookup.worker, atom::supervise_v, query_id, lookup.query,
                  std::move(actors),
                  caf::actor_cast<receiver_actor<atom::done>>(sink));
-      auto rp = self->make_response_promise<vast::uuid>();
+      auto rp = self->make_response_promise<partition_synopsis_pair>();
+      // TODO: Implement some kind of monadic composition instead of these
+      // nested requests.
       self
         ->request(sink, caf::infinite, atom::persist_v,
                   self->state.partition_path(new_partition_id),
                   self->state.partition_synopsis_path(new_partition_id))
         .then(
           [self, rp, old_partition_id, new_partition_id](
             std::shared_ptr<partition_synopsis>& synopsis) mutable {
+            // Make a copy so the caller and the meta index don't need
+            // synchronization.
+            auto result = partition_synopsis_pair{
+              new_partition_id,
+              std::make_shared<partition_synopsis>(*synopsis)};
+
```

Comment:
Why do you go to such great lengths to avoid the shared ref?

Reply:
It's going to be used by different actors, and thus may be accessed concurrently. (actually we should even copy this once per receiver, now that I think about it) It may be possible to hand out just a `shared_ptr<const partition_synopsis>` here and trust that concurrent reads will be safe, but currently our classes are not really designed for this, so we'd probably at least put a very prominent warning in the docs that implementations of `synopsis` need to be thread-safe.

Reply:
Are any synopses implementations using coders under the hood? Because that has a mutable cache, which I think would certainly be an issue. Other than that, the assumption in modern code should be that const equals thread safe access.

We could also make this a `caf::intrusive_cow_ptr`, which (1) does not need to be an unsafe message type, and (2) has exactly the interface you're looking for here. Accessing a non-const function would automatically clone if the synopsis is not unique.

Reply:
Sounds good, I'll test if that works.